src/packages/use-location-state/src/useQueryState/useQueryReducer.ts

```diff
@@ -62,13 +62,23 @@ export function useQueryReducer<
   )
   const activeQSI = queryStringInterface || hashQSI
 
-  // itemName & defaultValue is not allowed to be changed after init
-  const [defaultValue] = useState<ReducerState<R>>(() =>
+  // itemName is not allowed to be changed after init
+  const [defaultValue, setDefaultValue] = useState<ReducerState<R>>(() =>
     initStateFnOrOpts && typeof initStateFnOrOpts === 'function'
       ? initStateFnOrOpts(initialStateOrInitialArg as InitialArg)
       : (initialStateOrInitialArg as ReducerState<R>)
   )
 
+  // Update the default value if the initState/function changes
+  useEffect(() => {
+    const newDefaultValue =
+      initStateFnOrOpts && typeof initStateFnOrOpts === 'function'
+        ? initStateFnOrOpts(initialStateOrInitialArg as InitialArg)
+        : (initialStateOrInitialArg as ReducerState<R>)
+
+    setDefaultValue(newDefaultValue)
+  }, [initStateFnOrOpts, initialStateOrInitialArg])
```

Comment:
As described in my [comment](https://github.com/xiel/use-location-state/issues/345#issuecomment-1135805291), this would be breaking for people passing a init fn inline.

Closing the PR, we will keep the current behaviour (for now).

Still, thanks for the efforts to analyse the behaviour and open a PR.